### DISCUSSION TITLE
Increased specificity for mock database error reporting.

### DISCUSSION THREAD

**Discussion by jsimpson-gro** [SeaQL/sea-orm#1671] [General]
Having the mock database is one of my preferred features for `sea-orm`. Much appreciated.

When I was trying to get a unit test working, the mock layer was reporting an error as it tried to marshal the results provided to `append_query_results`:

```
.append_query_results([[maplit::btreemap! {
     "data_1_id" => Value::BigInt(Some(128)),
     "data_1_text" => Into::<Value>::into("apple")
}]])
```

into the result type required by `into_model`. (Yes, I am aware there are other ways of appending mock data, there is larger context that led to using a map, and the same problem is reproducible when using a `Model`.)

The error message was:

```
Error: Type Error: A null value was encountered while decoding "data_1_id"
```

I found this difficult to understand as there was non-null data present.

I believe the root of the problem is in the [`Value`](https://github.com/SeaQL/sea-query/blob/584d9eea9127e6e4e841ed9ef50a43f5dc19c856/src/value.rs#L337-L342) object, when an error occurs due to a type mismatch (the type in the destination struct is `String`, the type of the input data is `i32`). Probably that context is lost before the error is reported.

I like it that the error message indicates the name of the field. It would be nice the error message indicated invalid type (ideally even indicating the expected type), rather than null found.

Related, if the field is completely missing from the mock data, it also has the same error message (`A null value was encountered`), rather than indicating the field was not provided.

I've included a small project with a couple test cases that reproduces the problem as I experienced it.

[example-error-message.tar.gz](https://github.com/SeaQL/sea-orm/files/11569797/example-error-message.tar.gz)

**Comment by billy1624** [comment]
Hey @jsimpson-gro, you can get a details error log when you enabled the `debug-print` feature and setup a `tracing_subscriber`.

```rs
tracing_subscriber::fmt()
    .with_max_level(tracing::Level::DEBUG)
    .with_test_writer()
    .init();
```

```
---- executor::paginator::tests::num_pages stdout ----
2023-05-30T06:03:59.759568Z DEBUG sea_orm::driver::mock: SELECT COUNT(*) AS num_items FROM (SELECT "fruit"."id", "fruit"."name", "fruit"."cake_id" FROM "fruit") AS "sub_query"
2023-05-30T06:03:59.759811Z DEBUG sea_orm::executor::query: "Query Error: No column for ColIdx \"num_items\""
Error: Type("A null value was encountered while decoding \"num_items\"")
```

https://www.sea-ql.org/SeaORM/docs/install-and-config/debug-log/

**Reply by jsimpson-gro** [reply]
That certainly helps when difficult situations arrive, however a direct error message would be preferable. A good error message means seconds or minutes on a problem like this rather than hours or days. Even finding the incantation you provided is unnecessary effort for something that can occur very easily.
